There's something fishy:

```
2025-01-23T11:12:31.1455219Z [ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:generate-code (default) on project quarkus-integration-test-kafka-avro-apicurio2: Quarkus code generation phase has failed: InvocationTargetException: Failed to initialize application configuration: Configuration validation failed:
2025-01-23T11:12:31.1461049Z [ERROR] 	java.lang.IllegalArgumentException: SRCFG00013: No Converter registered for class io.quarkus.kafka.client.deployment.KafkaDevServicesBuildTimeConfig
2025-01-23T11:12:31.1462358Z [ERROR] -> [Help 1] .
```

Let me have a closer look.

--- a/extensions/kafka-client/deployment/pom.xml
+++ b/extensions/kafka-client/deployment/pom.xml
@@ -84,9 +84,6 @@
                                     <version>${project.version}</version>
                                 </path>
                             </annotationProcessorPaths>
-                            <compilerArgs>
-                                <arg>-AlegacyConfigRoot=true</arg>
-                            </compilerArgs>
                         </configuration>
                     </execution>
                 </executions>

--- a/extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java
+++ b/extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java
@@ -305,7 +305,7 @@ public class DevServicesKafkaProcessor {
     }
 
     private KafkaDevServiceCfg getConfiguration(KafkaBuildTimeConfig cfg) {
-        KafkaDevServicesBuildTimeConfig devServicesConfig = cfg.devservices;
+        KafkaDevServicesBuildTimeConfig devServicesConfig = cfg.devservices();
         return new KafkaDevServiceCfg(devServicesConfig);
     }
 
@@ -324,17 +324,17 @@ public class DevServicesKafkaProcessor {
         private final RedpandaBuildTimeConfig redpanda;
 
         public KafkaDevServiceCfg(KafkaDevServicesBuildTimeConfig config) {
-            this.devServicesEnabled = config.enabled.orElse(true);
-            this.provider = config.provider;
-            this.imageName = config.imageName.orElseGet(provider::getDefaultImageName);
-            this.fixedExposedPort = config.port.orElse(0);
-            this.shared = config.shared;
-            this.serviceName = config.serviceName;
-            this.topicPartitions = config.topicPartitions;
-            this.topicPartitionsTimeout = config.topicPartitionsTimeout;
-            this.containerEnv = config.containerEnv;
+            this.devServicesEnabled = config.enabled().orElse(true);
+            this.provider = config.provider();
+            this.imageName = config.imageName().orElseGet(provider::getDefaultImageName);
+            this.fixedExposedPort = config.port().orElse(0);
+            this.shared = config.shared();
+            this.serviceName = config.serviceName();
+            this.topicPartitions = config.topicPartitions();
+            this.topicPartitionsTimeout = config.topicPartitionsTimeout();
+            this.containerEnv = config.containerEnv();
 
-            this.redpanda = config.redpanda;
+            this.redpanda = config.redpanda();
         }
 
         @Override

--- a/extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaBuildTimeConfig.java
+++ b/extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaBuildTimeConfig.java
@@ -1,19 +1,23 @@
 package io.quarkus.kafka.client.deployment;
 
 import io.quarkus.runtime.annotations.ConfigDocSection;
-import io.quarkus.runtime.annotations.ConfigItem;
 import io.quarkus.runtime.annotations.ConfigPhase;
 import io.quarkus.runtime.annotations.ConfigRoot;
+import io.smallrye.config.ConfigMapping;
+import io.smallrye.config.WithDefault;
+import io.smallrye.config.WithName;
 
-@ConfigRoot(name = "kafka", phase = ConfigPhase.BUILD_TIME)
-public class KafkaBuildTimeConfig {
+@ConfigMapping(prefix = "quarkus.kafka")
+@ConfigRoot(phase = ConfigPhase.BUILD_TIME)
+public interface KafkaBuildTimeConfig {
     /**
      * Whether a health check is published in case the smallrye-health extension is present.
      * <p>
      * If you enable the health check, you must specify the `kafka.bootstrap.servers` property.
      */
-    @ConfigItem(name = "health.enabled", defaultValue = "false")
-    public boolean healthEnabled;
+    @WithName("health.enabled")
+    @WithDefault("false")
+    boolean healthEnabled();
 
     /**
      * Whether to enable Snappy in native mode.
@@ -21,24 +25,25 @@ public class KafkaBuildTimeConfig {
      * Note that Snappy requires GraalVM 21+ and embeds a native library in the native executable.
      * This library is unpacked and loaded when the application starts.
      */
-    @ConfigItem(name = "snappy.enabled", defaultValue = "false")
-    public boolean snappyEnabled;
+    @WithName("snappy.enabled")
+    @WithDefault("false")
+    boolean snappyEnabled();
 
     /**
      * Whether to load the Snappy native library from the shared classloader.
      * This setting is only used in tests if the tests are using different profiles, which would lead to
      * unsatisfied link errors when loading Snappy.
      */
-    @ConfigItem(name = "snappy.load-from-shared-classloader", defaultValue = "false")
-    public boolean snappyLoadFromSharedClassLoader;
+    @WithName("snappy.load-from-shared-classloader")
+    @WithDefault("false")
+    boolean snappyLoadFromSharedClassLoader();
 
     /**
      * Dev Services.
      * <p>
      * Dev Services allows Quarkus to automatically start Kafka in dev and test mode.
      */
-    @ConfigItem
     @ConfigDocSection(generated = true)
-    public KafkaDevServicesBuildTimeConfig devservices;
+    KafkaDevServicesBuildTimeConfig devservices();
 
 }

--- a/extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaDevServicesBuildTimeConfig.java
+++ b/extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaDevServicesBuildTimeConfig.java
@@ -5,11 +5,9 @@ import java.util.Map;
 import java.util.Optional;
 
 import io.quarkus.runtime.annotations.ConfigDocMapKey;
-import io.quarkus.runtime.annotations.ConfigGroup;
-import io.quarkus.runtime.annotations.ConfigItem;
+import io.smallrye.config.WithDefault;
 
-@ConfigGroup
-public class KafkaDevServicesBuildTimeConfig {
+public interface KafkaDevServicesBuildTimeConfig {
 
     /**
      * If Dev Services for Kafka has been explicitly enabled or disabled. Dev Services are generally enabled
@@ -17,16 +15,14 @@ public class KafkaDevServicesBuildTimeConfig {
      * {@code kafka.bootstrap.servers} is set or if all the Reactive Messaging Kafka channel are configured with a
      * {@code bootstrap.servers}.
      */
-    @ConfigItem
-    public Optional<Boolean> enabled = Optional.empty();
+    Optional<Boolean> enabled();
 
     /**
      * Optional fixed port the dev service will listen to.
      * <p>
      * If not defined, the port will be chosen randomly.
      */
-    @ConfigItem
-    public Optional<Integer> port;
+    Optional<Integer> port();
 
     /**
      * Kafka dev service container type.
@@ -44,10 +40,10 @@ public class KafkaDevServicesBuildTimeConfig {
      * <p>
      * Note that Strimzi and Kafka Native images are launched in Kraft mode.
      */
-    @ConfigItem(defaultValue = "redpanda")
-    public Provider provider = Provider.REDPANDA;
+    @WithDefault("redpanda")
+    Provider provider();
 
-    public enum Provider {
+    enum Provider {
         REDPANDA("docker.io/redpandadata/redpanda:v24.1.2"),
         STRIMZI("quay.io/strimzi-test-container/test-container:latest-kafka-3.7.0"),
         KAFKA_NATIVE("quay.io/ogunalp/kafka-native:latest");
@@ -68,8 +64,7 @@ public class KafkaDevServicesBuildTimeConfig {
      * <p>
      * Dependent on the provider.
      */
-    @ConfigItem
-    public Optional<String> imageName;
+    Optional<String> imageName();
 
     /**
      * Indicates if the Kafka broker managed by Quarkus Dev Services is shared.
@@ -82,8 +77,8 @@ public class KafkaDevServicesBuildTimeConfig {
      * <p>
      * Container sharing is only used in dev mode.
      */
-    @ConfigItem(defaultValue = "true")
-    public boolean shared;
+    @WithDefault("true")
+    boolean shared();
 
     /**
      * The value of the {@code quarkus-dev-service-kafka} label attached to the started container.
@@ -95,8 +90,8 @@ public class KafkaDevServicesBuildTimeConfig {
      * <p>
      * This property is used when you need multiple shared Kafka brokers.
      */
-    @ConfigItem(defaultValue = "kafka")
-    public String serviceName;
+    @WithDefault("kafka")
+    String serviceName();
 
     /**
      * The topic-partition pairs to create in the Dev Services Kafka broker.
@@ -106,29 +101,26 @@ public class KafkaDevServicesBuildTimeConfig {
      * <p>
      * The topic creation will not try to re-partition existing topics with different number of partitions.
      */
-    @ConfigItem
     @ConfigDocMapKey("topic-name")
-    public Map<String, Integer> topicPartitions;
+    Map<String, Integer> topicPartitions();
 
     /**
      * Timeout for admin client calls used in topic creation.
      * <p>
      * Defaults to 2 seconds.
      */
-    @ConfigItem(defaultValue = "2S")
-    public Duration topicPartitionsTimeout;
+    @WithDefault("2S")
+    Duration topicPartitionsTimeout();
 
     /**
      * Environment variables that are passed to the container.
      */
-    @ConfigItem
     @ConfigDocMapKey("environment-variable-name")
-    public Map<String, String> containerEnv;
+    Map<String, String> containerEnv();
 
     /**
      * Allows configuring the Redpanda broker.
      */
-    @ConfigItem
-    public RedpandaBuildTimeConfig redpanda;
+    RedpandaBuildTimeConfig redpanda();
 
 }

--- a/extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java
+++ b/extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java
@@ -325,7 +325,7 @@ public class KafkaProcessor {
     void loadSnappyIfEnabled(LaunchModeBuildItem launch, SnappyRecorder recorder, KafkaBuildTimeConfig config) {
         boolean loadFromSharedClassLoader = false;
         if (launch.isTest()) {
-            loadFromSharedClassLoader = config.snappyLoadFromSharedClassLoader;
+            loadFromSharedClassLoader = config.snappyLoadFromSharedClassLoader();
         }
         recorder.loadSnappy(loadFromSharedClassLoader);
     }
@@ -482,7 +482,7 @@ public class KafkaProcessor {
     @BuildStep
     HealthBuildItem addHealthCheck(KafkaBuildTimeConfig buildTimeConfig) {
         return new HealthBuildItem("io.quarkus.kafka.client.health.KafkaHealthCheck",
-                buildTimeConfig.healthEnabled);
+                buildTimeConfig.healthEnabled());
     }
 
     @BuildStep
@@ -546,7 +546,7 @@ public class KafkaProcessor {
 
         @Override
         public boolean getAsBoolean() {
-            return QuarkusClassLoader.isClassPresentAtRuntime("org.xerial.snappy.OSInfo") && config.snappyEnabled;
+            return QuarkusClassLoader.isClassPresentAtRuntime("org.xerial.snappy.OSInfo") && config.snappyEnabled();
         }
     }
 

--- a/extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/RedpandaBuildTimeConfig.java
+++ b/extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/RedpandaBuildTimeConfig.java
@@ -2,8 +2,7 @@ package io.quarkus.kafka.client.deployment;
 
 import java.util.Optional;
 
-import io.quarkus.runtime.annotations.ConfigGroup;
-import io.quarkus.runtime.annotations.ConfigItem;
+import io.smallrye.config.WithDefault;
 
 /**
  * Allows configuring the Redpanda broker.
@@ -11,8 +10,7 @@ import io.quarkus.runtime.annotations.ConfigItem;
  *
  * Find more info about Redpanda on <a href="https://redpanda.com/">https://redpanda.com/</a>.
  */
-@ConfigGroup
-public class RedpandaBuildTimeConfig {
+public interface RedpandaBuildTimeConfig {
 
     /**
      * Enables transaction support.
@@ -28,14 +26,13 @@ public class RedpandaBuildTimeConfig {
      * <a href="https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=89068820">KIP-360 (Improve reliability of
      * idempotent/transactional producer)</a> are <em>not</em> supported.
      */
-    @ConfigItem(defaultValue = "true")
-    public boolean transactionEnabled;
+    @WithDefault("true")
+    boolean transactionEnabled();
 
     /**
      * Port to access the Redpanda HTTP Proxy (<a href="https://docs.redpanda.com/current/develop/http-proxy/">pandaproxy</a>).
      * <p>
      * If not defined, the port will be chosen randomly.
      */
-    @ConfigItem
-    public Optional<Integer> proxyPort;
+    Optional<Integer> proxyPort();
 }

--- a/extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/RedpandaKafkaContainer.java
+++ b/extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/RedpandaKafkaContainer.java
@@ -60,7 +60,7 @@ final class RedpandaKafkaContainer extends GenericContainer<RedpandaKafkaContain
         command += String.format("--kafka-addr %s ", getKafkaAddresses());
         command += String.format("--advertise-kafka-addr %s ", getKafkaAdvertisedAddresses());
         command += "--set redpanda.auto_create_topics_enabled=true ";
-        if (redpandaConfig.transactionEnabled) {
+        if (redpandaConfig.transactionEnabled()) {
             command += "--set redpanda.enable_idempotence=true ";
             command += "--set redpanda.enable_transactions=true ";
         }
@@ -104,8 +104,8 @@ final class RedpandaKafkaContainer extends GenericContainer<RedpandaKafkaContain
             addFixedExposedPort(fixedExposedPort, DevServicesKafkaProcessor.KAFKA_PORT);
         }
 
-        if (redpandaConfig.proxyPort.isPresent()) {
-            addFixedExposedPort(redpandaConfig.proxyPort.get(), PANDAPROXY_PORT);
+        if (redpandaConfig.proxyPort().isPresent()) {
+            addFixedExposedPort(redpandaConfig.proxyPort().get(), PANDAPROXY_PORT);
         } else {
             addExposedPort(PANDAPROXY_PORT);
         }

--- a/extensions/kafka-client/runtime/pom.xml
+++ b/extensions/kafka-client/runtime/pom.xml
@@ -101,9 +101,6 @@
                                     <version>${project.version}</version>
                                 </path>
                             </annotationProcessorPaths>
-                            <compilerArgs>
-                                <arg>-AlegacyConfigRoot=true</arg>
-                            </compilerArgs>
                         </configuration>
                     </execution>
                 </executions>

